examples/qualcomm/llama2/qaihub_runner/io_memory.cpp

```diff
@@ -0,0 +1,477 @@
+  for (int i = 0, range = 1024 / thread_pool_.num_workers();
+       i < thread_pool_.num_workers();
+       ++i) {
+    lr_update_kv_.push_back(
+        {.start = i * range, .end = (i + 1) * range, .step = 1});
+  }
+}
+
+void KVCachedMemory::update_io(
+    int64_t cur_token,
+    int64_t pos,
+    std::vector<std::vector<Tensor>>& output_tensors) {
+  IO* ptr = static_cast<IO*>(data_ptr_.get());
+  int seq_len = 1023;
+  // update input_ids
+  ptr->input_ids = static_cast<int32_t>(cur_token);
+  // update causal mask for next token
+  ptr->attention_mask[seq_len - pos] = 65535;
+  // update position_ids
+  position_ids_cos_->set_data(position_ids_cos_->mutable_data<uint16_t>() + 64);
+  position_ids_sin_->set_data(position_ids_sin_->mutable_data<uint16_t>() + 64);
+
+  auto update_kv = [&](void* arg) {
+    LoopRange* lr = static_cast<LoopRange*>(arg);
+    // update v_cache
+    for (int i = lr->start; i < lr->end; i += lr->step) {
+      v_cache_in_[i]->set_data(v_cache_in_[i]->mutable_data<uint8_t>() + 128);
+      v_cache_out_[i]->set_data(v_cache_out_[i]->mutable_data<uint8_t>() + 128);
+    }
+    // update output tensors of v_cache, 256 is the number of kvs per shard
+    int shard = lr->start >> 8, offset = shard << 8;
+    int start = lr->start - offset, end = lr->end - offset;
+    for (int i = start; i < end; i += 32) {
+      for (int j = 0; j < 2; ++j) {
+        for (int k = 0; k < 32; ++k) {
+          // k, v are placed interleaved
+          int index = (i << 1) + (j << 5) + k;
+          ET_CHECK_MSG(
+              modules_[shard]->set_output_data_ptr(
+                  output_tensors[shard][index], index) == Error::Ok,
+              "failed to set output tensor for module %d's %d'th output "
+              "while updating kv_cache output tensors",
+              shard,
+              index);
+        }
+      }
+    }
+  };
+  for (int i = 0; i < lr_update_kv_.size(); ++i) {
+    futures_[i] = std::move(thread_pool_.issue(update_kv, &lr_update_kv_[i]));
+  }
+  // update k_cache by single thread, this part is cpu cache sensitive
+  for (int i = 0; i < k_cache_in_.size(); ++i) {
+    uint8_t* ptr_in = k_cache_in_[i]->mutable_data<uint8_t>();
+    const uint8_t* ptr_out = k_cache_out_[i]->data<uint8_t>();
+    for (size_t j = 0, offset = seq_len; j < 128; ++j, offset += seq_len) {
+      ptr_in[offset] = ptr_out[j];
+    }
+    k_cache_in_[i]->set_data(ptr_in + 1);
+  }
+  for (auto& future : futures_) {
+    future.wait();
+  }
+}
+
+ThreadPool::ThreadPool() : quit_(false) {
+  size_t hc = (std::thread::hardware_concurrency() + 3) / 4;
+  // maximum number should be divisible by head dimension which equals to 32
+  num_workers_ = min(32, hc * 4);
+  for (size_t i = 0; i < num_workers_; ++i) {
+    threads_.emplace_back([this]() {
+      while (1) {
+        std::unique_lock<std::mutex> lock(mutex_);
+        cv_.wait(lock, [this] { return !jobs_.empty() || quit_; });
+
+        if (quit_ && jobs_.empty())
```

Comment:
not sure on the semantics of quit_ here; I would have guessed that `quit_` would mean to stop even if not all the jobs are done

Reply:
Change to `stop_`.